scheduler/reconcile.go

```diff
@@ -0,0 +1,559 @@
+			canaries = nil
+		}
+	}
+
+	// Determine what set of alloations are on tainted nodes
+	untainted, migrate, lost := all.filterByTainted(a.taintedNodes)
+
+	// Create a structure for choosing names. Seed with the taken names which is
+	// the union of untainted and migrating nodes (includes canaries)
+	nameIndex := newAllocNameIndex(a.jobID, group, tg.Count, untainted.union(migrate))
+
+	// Stop any unneeded allocations and update the untainted set to not
+	// included stopped allocations.
+	canaryState := dstate != nil && dstate.DesiredCanaries != 0 && !dstate.Promoted
+	stop := a.computeStop(tg, nameIndex, untainted, migrate, lost, canaries, canaryState)
+	desiredChanges.Stop += uint64(len(stop))
+	untainted = untainted.difference(stop)
+
+	// Having stopped un-needed allocations, append the canaries to the existing
+	// set of untainted because they are promoted. This will cause them to be
+	// treated like non-canaries
+	if !canaryState {
+		untainted = untainted.union(canaries)
+		nameIndex.Set(canaries)
+	}
+
+	// Do inplace upgrades where possible and capture the set of upgrades that
+	// need to be done destructively.
+	ignore, inplace, destructive := a.computeUpdates(tg, untainted)
+	desiredChanges.Ignore += uint64(len(ignore))
+	desiredChanges.InPlaceUpdate += uint64(len(inplace))
+	if !existingDeployment {
+		dstate.DesiredTotal += len(destructive) + len(inplace)
+	}
+
+	// The fact that we have destructive updates and have less canaries than is
+	// desired means we need to create canaries
+	numDestructive := len(destructive)
+	strategy := tg.Update
+	requireCanary := numDestructive != 0 && strategy != nil && len(canaries) < strategy.Canary
+	if requireCanary && !a.deploymentPaused && !a.deploymentFailed {
+		number := strategy.Canary - len(canaries)
+		number = helper.IntMin(numDestructive, number)
+		desiredChanges.Canary += uint64(number)
+		if !existingDeployment {
+			dstate.DesiredCanaries = strategy.Canary
+		}
+
+		for _, name := range nameIndex.NextCanaries(uint(number), canaries, destructive) {
+			a.result.place = append(a.result.place, allocPlaceResult{
+				name:      name,
+				canary:    true,
+				taskGroup: tg,
+			})
+		}
+	}
+
+	// Determine how many we can place
+	canaryState = dstate != nil && dstate.DesiredCanaries != 0 && !dstate.Promoted
+	limit := a.computeLimit(tg, untainted, destructive, canaryState)
+
+	// Place if:
+	// * The deployment is not paused or failed
+	// * Not placing any canaries
+	// * If there are any canaries that they have been promoted
+	place := a.computePlacements(tg, nameIndex, untainted, migrate)
+	if !existingDeployment {
+		dstate.DesiredTotal += len(place)
+	}
+
+	if !a.deploymentPaused && !a.deploymentFailed && !canaryState {
+		// Place all new allocations
+		desiredChanges.Place += uint64(len(place))
+		for _, p := range place {
+			a.result.place = append(a.result.place, p)
+		}
+
+		// Do all destructive updates
+		min := helper.IntMin(len(destructive), limit)
+		limit -= min
```

Comment:
It isn't used after this but it is used to calculate the number of allowed changes. See line above and in the range iteration on 347